### PR TITLE
[Proposal] add proposal of daemosnet update model

### PR DESCRIPTION
**Pull request by hxcGit**
Signed-off-by: hxcGit <houxc_mail@163.com>

/kind design

#### What this PR does / why we need it:
Optimize workload update model in cloud-edge collaboration scenarios

#### Which issue(s) this PR fixes:

#914 

#### Special notes for your reviewer:

/assign @rambohe-ch 

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

**Comment by rambohe-ch** [comment]
@hxcGit Very appreciate you for posting proposal to upgrade edge workload. and i will take a look soon.

**Comment by rambohe-ch** [comment]
@hxcGit please rebase the master branch and push --force this pull reqeust.

File: docs/proposals/20220718-workload-update-model.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+#### Story 1
+
+- In some node-independent scenarios, such as a user owning a node, for example, on a Tesla car, the owner wants to independently update his application, instead of being controlled by the cluster administrator. At this time, the user can complete the application update process through a two-step process.
+  1. Check whether the latest version of the application is available.
+  2. Confirm the execution of application update.
+
+#### Story 2
+
+- In a openyurt cluster, there are some edge nodes, and some node's status are `not-ready` due to the network, when the administrator updates the application version, the administrator expects that for the `ready` edge nodes, the application can be updated to the latest version smoothly; for the `not-ready` edge nodes, they do not block the update process of other nodes.
```
Comment (huangyuqi) [inline docs/proposals/20220718-workload-update-model.md:69]: Story 2 may not be a suitable case for the `OTA` updating model.

File: docs/proposals/20220718-workload-update-model.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+3. If update is available, users can choose to apply update. (This is only from the user's perspective)
+4. OTA module removes the selected pod from the current node.（This step is achieved by calling the Kubernetes interface）'
+5. Kubelet on this node will synchronize pods information from Kubernetes APIServer.
+6. Kubelet is responsible for creating the latest version of pod.
+
+<div align="center">
+  <img src="../img/process.png" width=40% title="OTA Process">
+</div>
+
+##### OTA API
```
Comment (huangyuqi) [inline docs/proposals/20220718-workload-update-model.md:94]: where can the user call this api from ? Node?
If the api called from node,  so why don't we operate the pod or podlist?

File: docs/proposals/20220718-workload-update-model.md (hunk 1 above)
Comment (hxcGit) [reply docs/proposals/20220718-workload-update-model.md:69]: Yes, this is for administrator controlled updates.

File: docs/proposals/20220718-workload-update-model.md (hunk 2 above)
Comment (hxcGit) [reply docs/proposals/20220718-workload-update-model.md:94]: Users should call these APIs on edge nodes. Perhaps we can discuss this at community meeting tomorrow.

**Comment by rambohe-ch** [comment]
@hxcGit yes, we need to call RESTFUL api on edge nodes to upgrade edge pods. maybe we can define the RESTFUL api as following:
1. get all pods on edge node
`GET /pods HTTP/1.1`

2. upgrade specified pod on edge node
```
POST /openyurt.io/v1/{ns}/{podname} HTTP/1.1
BODY: {
    operation: upgrade,
}
```

**Comment by kadisi** [comment]
> @hxcGit yes, we need to call RESTFUL api on edge nodes to upgrade edge pods. maybe we can define the RESTFUL api as following:
> 
> 1. get all pods on edge node
>    `GET /pods HTTP/1.1`
> 2. upgrade specified pod on edge node
> 
> ```
> POST /openyurt.io/v1/{ns}/{podname} HTTP/1.1
> BODY: {
>     operation: upgrade,
> }
> ```

The API for `upgrading, suspending, and resuming` the POD can be designed in accordance with the API for POD logs and EXEC
```
/api/v1/namespaces/{ns}/pods/{podname}/exec
/api/v1/namespaces/{ns}/pods/{podname}/logs
```

So I suggest the following format：

-  get all pods on edge node

```
GET /pods HTTP/1.1
```


-  upgrade specified pod on edge node

```
POST /openyurt.io/v1/namespaces/{ns}/pods/{podname}/upgrade  HTTP/1.1
```

- suspend specified pod on edge node

```
POST /openyurt.io/v1/namespaces/{ns}/pods/{podname}/suspend  HTTP/1.1
```

- resume specified pod on edge node

```
POST /openyurt.io/v1/namespaces/{ns}/pods/{podname}/resume  HTTP/1.1
```

**Comment by rambohe-ch** [comment]
> > @hxcGit yes, we need to call RESTFUL api on edge nodes to upgrade edge pods. maybe we can define the RESTFUL api as following:
> > 
> > 1. get all pods on edge node
> >    `GET /pods HTTP/1.1`
> > 2. upgrade specified pod on edge node
> > 
> > ```
> > POST /openyurt.io/v1/{ns}/{podname} HTTP/1.1
> > BODY: {
> >     operation: upgrade,
> > }
> > ```
> 
> The API for `upgrading, suspending, and resuming` the POD can be designed in accordance with the API for POD logs and EXEC
> 
> ```
> /api/v1/namespaces/{ns}/pods/{podname}/exec
> /api/v1/namespaces/{ns}/pods/{podname}/logs
> ```
> 
> So I suggest the following format：
> 
> * get all pods on edge node
> 
> ```
> GET /pods HTTP/1.1
> ```
> 
> * upgrade specified pod on edge node
> 
> ```
> POST /openyurt.io/v1/namespaces/{ns}/pods/{podname}/upgrade  HTTP/1.1
> ```
> 
> * suspend specified pod on edge node
> 
> ```
> POST /openyurt.io/v1/namespaces/{ns}/pods/{podname}/suspend  HTTP/1.1
> ```
> 
> * resume specified pod on edge node
> 
> ```
> POST /openyurt.io/v1/namespaces/{ns}/pods/{podname}/resume  HTTP/1.1
> ```

@kadisi Thanks for your suggestion. agree+1

**Comment by rambohe-ch** [comment]
@hxcGit beside `OnDelete` update strategy, we need to add a new annotation(`apps.openyurt.io/ota: true`) for DaemonSet to specify `OTA` update strategy.

**Comment by hxcGit** [comment]
> > @hxcGit yes, we need to call RESTFUL api on edge nodes to upgrade edge pods. maybe we can define the RESTFUL api as following:
> > 
> > 1. get all pods on edge node
> >    `GET /pods HTTP/1.1`
> > 2. upgrade specified pod on edge node
> > 
> > ```
> > POST /openyurt.io/v1/{ns}/{podname} HTTP/1.1
> > BODY: {
> >     operation: upgrade,
> > }
> > ```
> 
> The API for `upgrading, suspending, and resuming` the POD can be designed in accordance with the API for POD logs and EXEC
> 
> ```
> /api/v1/namespaces/{ns}/pods/{podname}/exec
> /api/v1/namespaces/{ns}/pods/{podname}/logs
> ```
> 
> So I suggest the following format：
> 
> * get all pods on edge node
> 
> ```
> GET /pods HTTP/1.1
> ```
> 
> * upgrade specified pod on edge node
> 
> ```
> POST /openyurt.io/v1/namespaces/{ns}/pods/{podname}/upgrade  HTTP/1.1
> ```
> 
> * suspend specified pod on edge node
> 
> ```
> POST /openyurt.io/v1/namespaces/{ns}/pods/{podname}/suspend  HTTP/1.1
> ```
> 
> * resume specified pod on edge node
> 
> ```
> POST /openyurt.io/v1/namespaces/{ns}/pods/{podname}/resume  HTTP/1.1
> ```

Many thanks to your suggestion! I will use this form of APIs.

**Comment by hxcGit** [comment]
> @hxcGit beside `OnDelete` update strategy, we need to add a new annotation(`apps.openyurt.io/ota: true`) for DaemonSet to specify `OTA` update strategy.

OK, a nice suggestion!

**Comment by rambohe-ch** [comment]
@hxcGit I have a question, how to select one of the ways(OTA-update and NewController) for daemonset by end users?

**Comment by kadisi** [comment]
> @rambohe-ch I think we can use annotations apps.openyurt.io/upgrade-strategy: ota(for OTA-update), apps.openyurt.io/upgrade-strategy: immediate(for NewController) to distinguish them.

why use  `apps.openyurt.io/upgrade-strategy: immediate` ?

File: docs/proposals/20220718-workload-update-model.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+</div>
+
+##### Resources of interest
+
+- Daemonset
+- Node
+
+##### Steps
+
+**Pre-requisites**: Set application's updateStrategy to `type: OnDelete`; add a new annotation(`apps.openyurt.io/upgrade-strategy: auto`) for DaemonSet
```
Comment (kadisi) [inline docs/proposals/20220718-workload-update-model.md:121]: What will be the impact if `type: OnDelete` is not set? Do you need Webhook verification？

**Review by kadisi — COMMENTED** [review]
just some nits

**Comment by rambohe-ch** [comment]
> > @rambohe-ch I think we can use annotations apps.openyurt.io/upgrade-strategy: ota(for OTA-update), apps.openyurt.io/upgrade-strategy: immediate(for NewController) to distinguish them.
> 
> why use `apps.openyurt.io/upgrade-strategy: immediate` ?

@hxcGit @kadisi maybe `apps.openyurt.io/upgrade-strategy: auto` is more reasonable.

File: docs/proposals/20220718-workload-update-model.md (hunk 3 above)
Comment (rambohe-ch) [reply docs/proposals/20220718-workload-update-model.md:121]: @kadisi I agree with you that it's better to add validating webhook for upgrade strategy setting, but we can add webhook when coding, so i will merge this pull request at first.

**Comment by kadisi** [comment]
> > > @rambohe-ch I think we can use annotations apps.openyurt.io/upgrade-strategy: ota(for OTA-update), apps.openyurt.io/upgrade-strategy: immediate(for NewController) to distinguish them.
> > 
> > 
> > why use `apps.openyurt.io/upgrade-strategy: immediate` ?
> 
> @hxcGit @kadisi maybe `apps.openyurt.io/upgrade-strategy: auto` is more reasonable.

if ota , set  `apps.openyurt.io/upgrade-strategy: ota` annotation,   Otherwise, I don't set it？

**Comment by kadisi** [comment]
> > > @rambohe-ch I think we can use annotations apps.openyurt.io/upgrade-strategy: ota(for OTA-update), apps.openyurt.io/upgrade-strategy: immediate(for NewController) to distinguish them.
> > 
> > 
> > why use `apps.openyurt.io/upgrade-strategy: immediate` ?
> 
> @hxcGit @kadisi maybe `apps.openyurt.io/upgrade-strategy: auto` is more reasonable.

agree : `apps.openyurt.io/upgrade-strategy: auto`

**Review by rambohe-ch — APPROVED** [review]

**Pull request merged**